File: src/main/resources/eclairjs/sql/Encoder.js
/*                                                                         
* Copyright 2016 IBM Corp.                                                 
*                                                                          
* Licensed under the Apache License, Version 2.0 (the "License");          
* you may not use this file except in compliance with the License.         
* You may obtain a copy of the License at                                  
*                                                                          
*      http://www.apache.org/licenses/LICENSE-2.0                          
*                                                                          
* Unless required by applicable law or agreed to in writing, software      
* distributed under the License is distributed on an "AS IS" BASIS,        
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
* See the License for the specific language governing permissions and      
* limitations under the License.                                           
*/ 

(function () {

    //var JavaWrapper = require(EclairJS_Globals.NAMESPACE + '/JavaWrapper');
    //var Logger = require(EclairJS_Globals.NAMESPACE + '/Logger');
    //var Utils = require(EclairJS_Globals.NAMESPACE + '/Utils');
    //var logger = Logger.getLogger("sql.Encoder_js");


    var Encoder = Java.type('org.eclairjs.nashorn.wrap.sql.Encoder');
    
    /**
     * @classdesc
     * :: Experimental ::
     * Used to convert a JVM object of type `T` to and from the internal Spark SQL representation.
     *
     * == Scala ==
     * Encoders are generally created automatically through implicits from a `SparkSession`, or can be
     * explicitly created by calling static methods on {@link Encoders}.
     *
     * @example 
     *   import spark.implicits._
     *
     *   val ds = Seq(1, 2, 3).toDS() // implicitly provided (spark.implicits.newIntEncoder)
     *  
     *
     * == Java ==
     * Encoders are specified by calling static methods on {@link Encoders}.
     *
     * @example 
     *   List<String> data = Arrays.asList("abc", "abc", "xyz");
     *   Dataset<String> ds = context.createDataset(data, Encoders.STRING());
     *  
     *
     * Encoders can be composed into tuples:
     *
     * @example 
     *   Encoder<Tuple2<Integer, String>> encoder2 = Encoders.tuple(Encoders.INT(), Encoders.STRING());
     *   List<Tuple2<Integer, String>> data2 = Arrays.asList(new scala.Tuple2(1, "a");
     *   Dataset<Tuple2<Integer, String>> ds2 = context.createDataset(data2, encoder2);
     *  
     *
     * Or constructed from Java Beans:
     *
     * @example 
     *   Encoders.bean(MyClass.class);
     *  
     *
     * == Implementation ==
     *  - Encoders are not required to be thread-safe and thus they do not need to use locks to guard
     *    against concurrent access if they reuse internal buffers to improve performance.
     *
     * @since EclairJS 0.5 Spark  1.6.0
     * @class
     * @memberof module:eclairjs/sql
     */
    
    //
    //var Encoder = function() {
    //	 throw "Can't instantiate abstract class - Encoder";
    //\\ 	 this.logger = Logger.getLogger("Encoder_js");
    //\\ 	 JavaWrapper.call(this, jvmObject);
    //
    //};
    //
    //Encoder.prototype = Object.create(JavaWrapper.prototype);
    //
    //Encoder.prototype.constructor = Encoder;
    //
    
    
    /**
     *  Returns the schema of encoding this type of object as a Row. 
     * @returns {module:eclairjs/sql/types.StructType} 
     */
    //Encoder.prototype.schema = function() {
    //throw "not implemented by ElairJS";
    ////   var javaObject =  this.getJavaObject().schema();
    ////   return new StructType(javaObject);
    //};
    
    
    /**
     *  A ClassTag that can be used to construct and Array to contain a collection of `T`. 
     * @returns {ClassTag} 
     */
    //Encoder.prototype.clsTag = function() {
    //throw "not implemented by ElairJS";
    ////   var javaObject =  this.getJavaObject().clsTag();
    ////   return new ClassTag(javaObject);
    //};
    
    module.exports = Encoder;
})();

File: server/src/main/java/org/eclairjs/nashorn/wrap/sql/Encoder.java
package org.eclairjs.nashorn.wrap.sql;
/*                                                                         
* Copyright 2016 IBM Corp.                                                 
*                                                                          
* Licensed under the Apache License, Version 2.0 (the "License");          
* you may not use this file except in compliance with the License.         
* You may obtain a copy of the License at                                  
*                                                                          
*      http://www.apache.org/licenses/LICENSE-2.0                          
*                                                                          
* Unless required by applicable law or agreed to in writing, software      
* distributed under the License is distributed on an "AS IS" BASIS,        
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
* See the License for the specific language governing permissions and      
* limitations under the License.                                           
*/ 

import org.eclairjs.nashorn.Utils;
import org.eclairjs.nashorn.wrap.WrappedFunction;
import org.apache.log4j.Logger;
import org.eclairjs.nashorn.wrap.WrappedClass;


public class Encoder extends WrappedClass {

 static Logger logger = Logger.getLogger(Encoder.class);

    static WrappedFunction F_schema = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            logger.debug("schema");
            Object returnValue = null;
            org.apache.spark.sql.Encoder _encoder = (org.apache.spark.sql.Encoder) ((Encoder) thiz).getJavaObject();
            returnValue = _encoder.schema();
            return Utils.javaToJs(returnValue);
            //return new org.eclairjs.nashorn.wrap.sql.types.StructType((org.apache.spark.sql.types.StructType)returnValue);
        }
    };

    static WrappedFunction F_clsTag = new WrappedFunction() {
        @Override
        public Object call(Object thiz, Object... args) {
            logger.debug("clsTag");
            Object returnValue = null;
            org.apache.spark.sql.Encoder _encoder = (org.apache.spark.sql.Encoder) ((Encoder) thiz).getJavaObject();
            returnValue = _encoder.clsTag();
            return Utils.javaToJs(returnValue);
            //return new ClassTag((ClassTag)returnValue);
        }
    };


    private org.apache.spark.sql.Encoder _encoder;

    public Encoder(org.apache.spark.sql.Encoder _encoder)
    {
       logger.debug("constructor");
       this._encoder = _encoder;
    }

    static public String getModuleName() {
        return "sql.Encoder";
    }

    public boolean checkInstance(Object other) {
        return other instanceof Encoder;
    }

    public Object getJavaObject() {
        return _encoder;
    }

    @Override
    public String toString() {

        return _encoder.toString();
    }

    public String getClassName() {
        return "Encoder";
    }

    // get the value of that named property
    @Override
    public Object getMember(String name) {
        switch (name) {
            case "schema":
                return F_schema;
            case "clsTag":
                return F_clsTag;
        }
        return super.getMember(name);
    }

    @Override
    public boolean hasMember(String name) {
        switch (name) {
            case "schema":
            case "clsTag":
                return true;
        }
        return super.hasMember(name);
    }

}
